### PR TITLE
add icons for 'execute current chunk'

### PR DESCRIPTION
**Pull request by kevinushey**
NOTE: There's a lot of black magic in this PR, so be warned... (there could be cleaner ways of implementing this, but I didn't have much luck in discovering them)

chunk-icons

This PR adds 'run' icons to chunks that, when clicked, execute that chunk.

## TODO

- [ ] Enable only for R code chunks
- [ ] Consider enabling for Rcpp chunks (would call 'sourceCpp' on the contents)
- [ ] Ensure Sweave highlighter changes don't regress anywhere
- [ ] Consider better icons + styling

### CODE REVIEW

File: src/gwt/acesupport/acemode/sweave_background_highlighter.js [hunk 1]
```diff
@@ -136,29 +136,58 @@ define("mode/sweave_background_highlighter", function(require, exports, module)
...
+         // If the initial state is an 'end' state, we need to update
+         // all markers later on in the document, as we may have just
+         // closed a chunk.
+         var state = this.$rowState[startRow];
+         if (state === TYPE_END)
+            endRow = this.$doc.getLength() - 1;
+
          for (var row = startRow; row <= endRow; row++) {
-            var foreign = this.$rowState[row] != TYPE_TEXT;
-            if (!!foreign != !!this.$markers[row]) {
```
Comment (kevinushey) [inline src/gwt/acesupport/acemode/sweave_background_highlighter.js]: I don't know what this line is doing (too clever for its own good?) and was encountering problems with identifying the 'start' of a chunk so I ended up re-writing this block of code.

**Comment by jjallaire** [comment]
Great! We'll review and merge once we get the tree back open.

This is going to be a very high impact feature :-)

**Comment by kevinushey** [comment]
How do you feel about having a mini-toolbar for each chunk, with icons for navigation as well? E.g.

- Go to first chunk
- Go to previous chunk
- Go to next chunk
- Go to last chunk
- Evaluate this chunk

Or would that be overkill?

**Comment by jjallaire** [comment]
I think some type of mini toolbar would be awesome. We could do some chink
navigation as you suggest (but we don't want it to get too busy with
buttons). We could also use it as a way to expose common chunk options
(e.g. echo).

If we have the basic infrastructure we can work with Paul on the look and
having just enough functionality to be useful without being distracting.

On May 7, 2015, at 8:29 PM, Kevin Ushey <notifications@github.com> wrote:

How do you feel about having a mini-toolbar for each chunk, with icons for
navigation as well? E.g.

   - Go to first chunk
   - Go to previous chunk
   - Go to next chunk
   - Go to last chunk
   - Evaluate this chunk

Or would that be overkill?

—
Reply to this email directly or view it on GitHub
<https://github.com/rstudio/rstudio/pull/415#issuecomment-100054482>.

**Comment by kevinushey** [comment]
New treatment with just the plain green arrow:

screen shot 2015-05-18 at 1 49 30 pm

With this, I think it's okay to merge. Note that the icons will remain visible for non-R code chunks (since figuring that out in the Sweave highlighter would tangle code up a bit) and because the UI is lightweight enough that it's not a huge deal IMO (at least at this point).

**Comment by jjallaire** [comment]
It would be awesome if we had a tooltip saying "Run Current Chunk (keyboard-shortcut)" for the button. Is that relatively straightforward?

**Comment by kevinushey** [comment]
Just added that here: https://github.com/rstudio/rstudio/commit/c59728a3ab4589bdaa0b4fbb31f41a975bd5b199

**Comment by jjallaire** [comment]
Awesome, thanks!

**Pull request merged**